include/stdexec/__detail/__basic_sender.hpp

```diff
@@ -73,10 +77,239 @@ namespace stdexec {
     STDEXEC_PRAGMA_POP()
 
     struct __tie {
-      template <class _Tag, class _Data, class... _Children>
-      constexpr auto operator()(_Tag, _Data&& __data, _Children&&... __children) const noexcept {
-        return std::tuple<_Tag, _Data&&, _Children&&...>{
-          {}, (_Data&&) __data, (_Children&&) __children...};
+      template <class _Tag, class _Data, class... _Child>
+      STDEXEC_ATTRIBUTE((always_inline))
+      constexpr auto operator()(_Tag, _Data&& __data, _Child&&... __child) const noexcept {
+        return std::tuple<_Tag, _Data&&, _Child&&...>{{}, (_Data&&) __data, (_Child&&) __child...};
+      }
+    };
+
+    // Note: This is UB. UBSAN allows it for now.
+    template <class _Parent, class _Child>
+    _Parent* __parent_from_child(_Child* __child, _Child _Parent::*__mbr_ptr) noexcept {
+      alignas(_Parent) char __buf[sizeof(_Parent)];
+      _Parent* __parent = (_Parent*) &__buf;
+      const std::ptrdiff_t __offset = (char*) &(__parent->*__mbr_ptr) - __buf;
+      return (_Parent*) ((char*) __child - __offset);
+    }
+
+    template <class _Sexpr, class _Receiver>
+    struct __op_state;
+
+    template <class _Receiver, class _Sexpr, class _Idx>
+    struct __receiver {
+      using is_receiver = void;
+      using __sexpr = _Sexpr;
+      using __index = _Idx;
+
+      // A pointer to the parent operation state, which contains the one created with
+      // this receiver.
+      __op_state<_Sexpr, _Receiver>* __op_;
+
+      template <class _ChildSexpr, class _ChildReceiver>
+      static __receiver __from_op_state(__op_state<_ChildSexpr, _ChildReceiver>* __child) noexcept {
```

Comment:
I think that if we want to make this generally available in the library then it needs to be an official customisation point / protocol that user-defined operation-states/receivers can implement, and not be hidden behind stdlib ugly names. i.e. define a concept for `inlinable_receiver_in<R, Op>` that requires `receiver<r>` and `static R R::from_op_state(Op*) noexcept;`.

Then have a public `op_state_base<Derived, R>` base class that any user's op-state class can inherit from and get the benefit of op-state inlining.

Reply:
I don't want this to be a generally available customization point, esp considering that it relies on UB. At some point, compilers might start rejecting this and I have to abandon the effort. That's hard if it's a public interface with users.

Reply:
Can we find a way that would allow this to be not-UB?
e.g. by using base-classes to hold child operation-states for which we can validly cast from base-class (child) to derived class (parent)

Reply:
I don't know.

Reply:
I've been noodling on this a little - maybe something along the lines of this?
https://godbolt.org/z/MM59PTvK4